helix-term/src/commands.rs

```diff
@@ -3613,6 +3613,70 @@ pub fn completion(cx: &mut Context) {
     );
 }
 
+fn completion_sources(cx: &mut Context) {
+    let (view, doc) = current_ref!(cx.editor);
+    let text = doc.text().clone();
+    let cursor = doc.selection(view.id).primary().cursor(text.slice(..));
+    let callback = async move {
+        let trigger_offset = cursor;
+
+        // TODO: trigger_offset should be the cursor offset but we also need a starting offset from where we want to apply
+        // completion filtering. For example logger.te| should filter the initial suggestion list with "te".
+
+        use helix_core::chars;
+        let mut iter = text.chars_at(cursor);
+        iter.reverse();
```

Comment:
From what I'm able to gather, I think the intention of this is to find the chars that the user has typed so far by going backwards from the cursor. But according to the docs, [`chars_at`](https://docs.rs/ropey/latest/ropey/struct.Rope.html#method.chars_at)

>Creates an iterator over the chars of the Rope, starting at char char_idx.

If I'm understanding this right, if we have the text and cursor:

```
I can eat glas[s], it will not hurt me
```

Then `chars_at(cursor)` would be "it will not hurt me", and the reverse would be "em truh ton lliw ti". And so going from here, the offset would be 2, which would make the prefix `ass` instead of `glass`.

Also yeah, echoing archseer, repeating all the work of collecting words on every single idle timeout is going to be a problem. I think we'll need to do this a different way.